pkg/services/ngalert/state/manager.go

```diff
@@ -183,9 +183,7 @@ func (st *Manager) ProcessEvalResults(ctx context.Context, evaluatedAt time.Time
 		}
 	}
 
-	if st.historian != nil {
-		st.historian.RecordStates(ctx, alertRule, changedStates)
-	}
+	st.historian.RecordStates(ctx, alertRule, changedStates)
```

Comment:
This is a common problem of such patterns: here and below we do some work in order to prepare data for the historian (filtering). Although it is not optimized here (it should be), with this change we will not be able to do so, and some other optimizations.

Reply:
Sure, but this is a very tiny optimization - the original code also did not even take advantage of this in the first place. I'd be very surprised if this had a measurable impact on our current perf.

IMO, the benefits of bug-proofing and simplicity here outweigh this, and I feel that instead we could spend time simplifying the manager logic to handle this rather than go the other way.

If we feel that strongly about optimizing out this operation, I'd still rather simply push the operation below historian interface over discarding this.

Reply:
please check this PR https://github.com/grafana/grafana/pull/58358 it reduces the scope of usage of historian and get rid of the branching in ProcessEvalResults

Reply:
If you are rushing to get your PR in I can do that here and rebase off yours.

Reply:
Yes, as long as we move the filtering of the transition to historian we can replace it with the "Null object". However, it's a bit tricky: as you can see the history writing for stale states is different than for regular ones. Therefore, before we move filtering to the historian side, I think we should change\fix the filtering of stale states. After that the introduction of null object will be reasonable and simple